crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+        }
+    }
+}
+
+/// Possible statuses of an MPC session:
+/// - Active: The session is currently running; new messages will be forwarded to the session.
+/// - Finalizing: The session is finished and pending removal; incoming messages will not be forwarded,
+/// - Finished: The session removed from active instances; incoming messages will not be forwarded,
+/// but will not be marked as malicious.
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum MPCSessionStatus<Output> {
+    Active,
+    Finalizing(Output),
+    Finished(Output),
+}
+
+/// The `MPCService` is responsible for managing MPC instances:
+/// - keeping track of all MPC instances,
+/// - executing all active instances, and
+/// - (de)activating instances.
+pub struct SignatureMPCManager<P: CreatableParty> {
+    mpc_instances: HashMap<ObjectID, SignatureMPCInstance<P>>,
+    /// Used to keep track of the order in which pending instances are received so they are activated in order of arrival.
+    pending_instances_queue: VecDeque<ObjectID>,
+    // TODO (#257): Make sure the counter is always in sync with the number of active instances.
+    active_instances_counter: usize,
+    consensus_adapter: Arc<dyn SubmitToConsensus>,
+    pub epoch_store: Weak<AuthorityPerEpochStore>,
+    pub max_active_mpc_instances: usize,
+    auxiliary_input: P::AuxiliaryInput,
+    pub epoch_id: EpochId,
+    /// The total number of parties in the chain
+    /// We can calculate the threshold and parties IDs (indexes) from it
+    number_of_parties: usize,
+}
+
+/// Needed to be able to iterate over a vector of generic MPCInstances with Rayon
+unsafe impl<P: CreatableParty + Sync + Send> Send for SignatureMPCInstance<P> {}
+
+impl<P: CreatableParty + Sync + Send> SignatureMPCManager<P> {
+    pub fn new(
+        consensus_adapter: Arc<dyn SubmitToConsensus>,
+        epoch_store: Weak<AuthorityPerEpochStore>,
+        epoch_id: EpochId,
+        max_active_mpc_instances: usize,
+        number_of_parties: usize,
+        auxiliary_input: P::AuxiliaryInput,
+    ) -> Self {
+        Self {
+            mpc_instances: HashMap::new(),
+            pending_instances_queue: VecDeque::new(),
+            active_instances_counter: 0,
+            consensus_adapter,
+            epoch_store,
+            epoch_id,
+            max_active_mpc_instances,
+            // TODO (#268): Take into account the validator's voting power
+            auxiliary_input,
+            number_of_parties,
+        }
+    }
+
+    /// Tries to verify that the received output for the MPC session matches the one generated locally.
+    /// Returns true if the output is correct, false otherwise.
+    pub fn try_verify_output(
+        &self,
+        output: &Vec<u8>,
+        session_id: &ObjectID,
+    ) -> anyhow::Result<bool> {
+        let Some(instance) = self.mpc_instances.get(session_id) else {
+            return Ok(false);
+        };
+        let MPCSessionStatus::Finalizing(stored_output) = &instance.status else {
```

Comment:
@ItayLevyOfficial please address this.

Reply:
@dolevmu The threshold decryption is indeed a special case and we'll handle it when we'll implement the sign flow. 

@ycscaly Regarding _So let's assume we somehow didn't participate in the final round_, AFAIU, since all the transactions as of now are being processed synchronously, at the same order by all the validators, the only situation in which a validator may not "participate" in the last round is if it's still syncing with the chain. In this you're right - the validators should only receive message but take no active actions, such as marking a specific output as malicious. I've created an issue & added a TODO for this case.

@cdesaintguilhem let's discuss your suggestions on our next call